statesync/syncer.go

```diff
@@ -109,17 +119,23 @@ func (s *syncer) AddSnapshot(peer p2p.Peer, snapshot *snapshot) (bool, error) {
 	return added, nil
 }
 
-// AddPeer adds a peer to the pool. For now we just keep it simple and send a single request
-// to discover snapshots, later we may want to do retries and stuff.
-func (s *syncer) AddPeer(peer p2p.Peer) {
-	s.logger.Debug("Requesting snapshots from peer", "peer", peer.ID())
-	peer.Send(SnapshotChannel, mustEncodeMsg(&ssproto.SnapshotsRequest{}))
+// AddPeer adds a peer to the pool. For now we just keep it simple and send a
+// single request to discover snapshots, later we may want to do retries and stuff.
+func (s *syncer) AddPeer(peer p2p.PeerID) {
+	s.logger.Debug("Requesting snapshots from peer", "peer", peer.String())
+
+	go func() {
+		s.snapshotCh <- p2p.Envelope{
```

Comment:
these one channel blocking writes in a separate goroutine scare me 😨

Reply:
AFAIK, the old legacy code, `Peer#Send` is async and thus non-blocking? Correct me if I'm wrong. If so, this mimics the same behavior.

So either we:

1. Assume this is buffered
2. Keep as-is
3. Block and make it synchronous
3. Allow envelopes to be dropped via a `select`

Reply:
Scheduling of sending should be the responsibility of the router (or temporarily, the shim). If we want to do parallel sends, then the router can handle that internally, and do stuff like bounding of sender pools and peer prioritization. Allowing reactor channels to block is a backpressure mechanism.

Reply:
Ok will go with option (3).